test/e2e/operator_test.go

```diff
@@ -1075,6 +1076,295 @@ func TestContainerLogging(t *testing.T) {
+	}
+	defer func() {
+		if err := kclient.Delete(context.TODO(), echoRoute); err != nil {
+			t.Fatalf("failed to delete route %s/%s: %v", echoRoute.Namespace, echoRoute.Name, err)
+		}
+	}()
+
+	// Create a client pod that sends a request to the echo route and checks
+	// whether it gets the expected number of X-Forwarded-For headers.
+	kubeConfig, err := config.GetConfig()
+	if err != nil {
+		t.Fatalf("failed to get kube config: %v", err)
+	}
+	client, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		t.Fatalf("failed to create kube client: %v", err)
+	}
+	var testPodCount int
+	testRoute := func(expectedMatches int, headers []string) {
+		t.Helper()
+		curlArgs := []string{
+			"-s",
+			"--retry", "15", "--retry-delay", "1",
+			"--max-time", "2",
+			"--resolve",
+			echoRoute.Spec.Host + ":80:" + service.Spec.ClusterIP,
+		}
+		for _, header := range headers {
+			curlArgs = append(curlArgs, "-H", header)
+		}
+		curlArgs = append(curlArgs, "http://"+echoRoute.Spec.Host)
+		testPodCount++
+		clientPod := &corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("forwardedheader%d", testPodCount),
+				Namespace: echoRoute.Namespace,
+			},
+			Spec: corev1.PodSpec{
+				Containers: []corev1.Container{
+					{
+						Name:    "curl",
+						Image:   deployment.Spec.Template.Spec.Containers[0].Image,
+						Command: []string{"/bin/curl"},
+						Args:    curlArgs,
+					},
+				},
+				RestartPolicy: corev1.RestartPolicyNever,
+			},
+		}
+		if err := kclient.Create(context.TODO(), clientPod); err != nil {
+			t.Fatalf("failed to create pod %s/%s: %v", clientPod.Namespace, clientPod.Name, err)
+		}
+		defer func() {
+			if err := kclient.Delete(context.TODO(), clientPod); err != nil {
+				if !errors.IsNotFound(err) {
+					t.Fatalf("failed to delete pod %s/%s: %v", clientPod.Namespace, clientPod.Name, err)
+				}
+			}
+		}()
+		err = wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+			readCloser, err := client.CoreV1().Pods(clientPod.Namespace).GetLogs(clientPod.Name, &corev1.PodLogOptions{
+				Container: "curl",
+				Follow:    false,
+			}).Stream(context.TODO())
+			if err != nil {
+				t.Logf("failed to read output from pod %s: %v", clientPod.Name, err)
+				return false, nil
+			}
+			scanner := bufio.NewScanner(readCloser)
+			defer func() {
+				if err := readCloser.Close(); err != nil {
+					t.Errorf("failed to close reader for pod %s: %v", clientPod.Name, err)
+				}
+			}()
+			var numMatches int
+			for scanner.Scan() {
+				line := scanner.Text()
+				if strings.HasPrefix(line, "x-forwarded-for:") {
+					numMatches++
+					t.Logf("found match %d of %d expected: %s", numMatches, expectedMatches, line)
+					if numMatches > expectedMatches {
+						return false, fmt.Errorf("got %d x-forwarded-for headers, expected %d", numMatches, expectedMatches)
+					}
+				}
+			}
+			return numMatches == expectedMatches, nil
```

Comment:
Line 1263 checks for the case where `numMatches > expectedMatches`. Perhaps before the return on line 1268, we should check for `numMatches > 0 && numMatches < expectedMatches` and return an error if that condition is true. This way, for example, the polling loop would half if only 1 x-forwarded-for header was found in the curl response when 2 were expected, and the test failure message would be more verbose. Thoughts? I may be overthinking this, but since you're calling `testRoute` several times I figured it would be worth mentioning.

Reply:
That makes sense.  It is unlikely that the logs will have partial headers, so returning an error if 0 &lt; `numMatches` &lt; `expectedMatches` should safe to do.